xdsl/transforms/jax_use_donated_arguments.py

```diff
@@ -0,0 +1,144 @@
+import itertools
+from collections.abc import Sequence
+from dataclasses import dataclass
+
+from xdsl.context import MLContext
+from xdsl.dialects import builtin
+from xdsl.dialects.bufferization import MaterializeInDestinationOp
+from xdsl.dialects.builtin import FunctionType, TensorType
+from xdsl.dialects.func import FuncOp, ReturnOp
+from xdsl.ir import BlockArgument, Operation, SSAValue
+from xdsl.irdl import VarOperand
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+def map_outputs_to_inputs(
+    donated_inputs: Sequence[BlockArgument], outputs: VarOperand
+) -> dict[int, int]:
+    """
+    Find suitable donated buffers for each of returned variables.
+    Each buffer can be used only once.
+    Types of the buffer and the variable should match.
+    """
+    used_inputs_idx: set[int] = set()
+    output_input_mapping: dict[int, int] = {}
+
+    for outputs_idx, output in enumerate(outputs):
+        if isinstance(output.owner, MaterializeInDestinationOp) and isinstance(
+            output.owner.dest, BlockArgument
+        ):
+            # the return value is already bufferized in a donated buffer
+            continue
```

Comment:
is this still necessary? With your changes, it should be safe to make the assumption that the result can be donated if the attribute is present.